app/gui/src/model/project/synchronized.rs

```diff
@@ -544,6 +544,7 @@ impl Project {
                 Event::Notification(Notification::ExecutionStatus(_)) => {}
                 Event::Notification(Notification::ExecutionComplete { context_id }) => {
                     execution_update_handler(context_id, ExecutionUpdate::Completed);
+                    publisher.notify(model::project::Notification::ExecutionFinished);
```

Comment:
If this notification is sent also when _any_ execution finishes (also those performed automatically after a change), then we could have a race where:
1. User did a change, the automatic execution starts
2. User clicks the play button. We set read-only flag to true.
3. We receive "ExecutionComplete" event about execution launched in 1. We turn off "read-only" flag.
4. Only after some time we receive "ExecutionComplete" event about execution launched in 2. Here we should turn off the flag instead.

Reply:
Good catch. But is the old execution not aborted when we request a new one? It could still happen in a very small time window, though. But I don't think we currently have a way to match a request for re-computation with its status update, unless I'm missing something.

Reply:
@4e6 could you help us with that? Will the engine stop the current execution when receiving a `recompute` message?

Reply:
Yes, when the `recompute` message is received, the engine aborts the current execution and re-schedules a new one. In general, the engine follows the principle that if it needs to schedule a new execution, it makes sure that the other execution is stopped. Otherwise, it may accumulate those commands and re-execute the program in a sequence.